dia-batching-server/src/api/binance.rs

```diff
@@ -0,0 +1,115 @@
+use crate::api::error::BinanceError;
+use crate::types::{AssetSpecifier, Quotation};
+use chrono::Utc;
+use rust_decimal::prelude::Zero;
+use rust_decimal::Decimal;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+
+pub struct BinancePriceApi {
+	client: BinanceClient,
+}
+
+impl BinancePriceApi {
+	pub fn new() -> Self {
+		let client: BinanceClient = BinanceClient::default();
+
+		Self { client }
+	}
+
+	pub async fn get_price(&self, asset: &AssetSpecifier) -> Result<Quotation, BinanceError> {
+		let binance_asset_identifier = Self::convert_to_binance_id(&asset);
+
+		match self.client.price(binance_asset_identifier).await {
+			Ok(price) => Ok(Quotation {
+				symbol: asset.symbol.clone(),
+				name: asset.symbol.clone(),
+				blockchain: Some(asset.blockchain.clone().into()),
+				price: price.price,
+				supply: Decimal::zero(),
+				time: Utc::now().timestamp().unsigned_abs(),
+			}),
+			Err(error) => {
+				log::warn!("Error getting price for {:?} from Binance: {:?}", asset, error);
+				Err(BinanceError(format!("Could not get price for {:?}", asset)))
+			},
+		}
+	}
+
+	// We assume here that the `symbol` field of the `AssetSpecifier` is a valid asset for Binance.
+	// The `blockchain` field is not used and ignored.
+	fn convert_to_binance_id(asset: &AssetSpecifier) -> String {
+		asset.symbol.to_uppercase()
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct BinancePrice {
+	pub symbol: String,
+	pub price: Decimal,
+}
+
+pub struct BinanceClient {
+	host: String,
+	inner: reqwest::Client,
+}
+
+impl BinanceClient {
+	pub fn default() -> Self {
+		Self::new("https://api.binance.com".to_string())
+	}
+
+	pub fn new(host: String) -> Self {
+		let inner = reqwest::Client::new();
+
+		Self { host, inner }
+	}
+
+	async fn get<R: DeserializeOwned>(&self, endpoint: &str) -> Result<R, BinanceError> {
+		let url = reqwest::Url::parse(
+			format!("{host}/{ep}", host = self.host.as_str(), ep = endpoint).as_str(),
+		)
+		.expect("Invalid URL");
+
+		let response = self
+			.inner
+			.get(url)
+			.send()
+			.await
+			.map_err(|e| BinanceError(format!("Failed to send request: {}", e.to_string())))?;
+
+		if !response.status().is_success() {
+			let result = response.text().await;
+			return Err(BinanceError(format!(
+				"Binance API error: {}",
+				result.unwrap_or("Unknown".to_string()).trim()
+			)));
+		}
+
+		let result = response.json().await;
```

Comment:
All the remaining fields (`highPrice`, `lowPrice`, etc.) are ignored when result is deserialized to `BinancePrice`?

Reply:
There is no `highPrice` or `lowPrice` on the response, see the documentation [here](https://developers.binance.com/docs/binance-spot-api-docs/rest-api/public-api-endpoints#symbol-price-ticker) 😅

Reply:
Well I saw it yesterday but perhaps was another endpoint ;). Now the links to the docs are not working. But whatever the case it's good!